src/lib/flash-helper.js

```diff
@@ -38,29 +38,55 @@ async function flashFiles({ device, flashSteps, resetAfterFlash = true, ui }) {
 }
 
 async function _flashDeviceInNormalMode(device, data, { name, progress } = {}) {
-	if (device.isInDfuMode) {
-		// put device in normal mode
-		if (progress) {
-			progress({ event: 'switch-mode', mode: 'normal' });
-		}
-		device = await usbUtils.reopenInNormalMode(device, { reset: true });
-	}
-	// put the device in listening mode to prevent cloud connection
-	try {
-		await device.enterListeningMode();
-	} catch (error) {
-		// ignore if the device is already in listening mode
-	}
-
 	// flash the file in normal mode
 	if (progress) {
 		progress({ event: 'flash-file', filename: name });
 	}
-	await device.updateFirmware(data, { progress, timeout: FLASH_TIMEOUT });
 
-	// wait for the device to apply the firmware
-	await delay(FLASH_APPLY_DELAY);
-	device = await usbUtils.reopenInNormalMode(device, { reset: false });
+	const start = Date.now();
+	while (Date.now() - start < FLASH_TIMEOUT) {
+		try {
+			device = await usbUtils.reopenInNormalMode(device);
+			await device.updateFirmware(data, { progress, timeout: FLASH_TIMEOUT });
+			return device;
+		} catch (error) {
+			// ignore error from attempts to flash to external flash
+		}
+	}
+	throw new Error('Unable to flash device');
+}
+
+async function prepareDeviceForFlash({ device, mode, progress }) {
+	if (device.isOpen) {
+		await device.close();
+	}
+	// check if open
+	device = await usbUtils.reopenDevice(device);
+	switch (mode) {
+		case 'normal':
+			if (device.isInDfuMode) {
+				// put device in normal mode
+				if (progress) {
+					progress({ event: 'switch-mode', mode: 'normal' });
+				}
+				device = await usbUtils.reopenInNormalMode(device, { reset: true });
+			}
+			try {
+				await device.enterListeningMode();
+				await delay(1000); // Just in case
```

Comment:
check this 👉  https://github.com/particle-iot/device-os-flash-util/blob/master/lib/usb.js#L64